### ISSUE TITLE
Add native support for TR-HASH deterministic token-routed MoE

### ISSUE THREAD

**Issue by Complexity-ML** [huggingface/transformers#48193]
### Model description

We are the model authors and would like native Transformers support for TR-HASH MoE so the public checkpoint can load with `AutoModelForCausalLM.from_pretrained(...)` without `trust_remote_code=True`.

TR-HASH is a 201.2M-parameter decoder-only causal language model with 16 layers, GQA (14 query heads / 2 KV heads), tied embeddings, four stored experts with top-2 activation, and an always-on shared SwiGLU path.

Unlike conventional sparse MoE models, expert selection is not produced by a learned router. It is a deterministic function of token IDs, the layer index, and persisted per-layer multi-hash routing tables.

The principal implementation differences are:

1. The MLP receives token IDs in addition to hidden states.
2. Routing tables are persisted integer buffers and must stay paired with expert weights.
3. Expert projections are stored as stacked tensors:
   - `expert_gate`: `[num_experts, hidden_size, expert_width]`
   - `expert_up`: `[num_experts, hidden_size, expert_width]`
   - `expert_down`: `[num_experts, expert_width, hidden_size]`
4. A shared SwiGLU branch is always active alongside the routed experts.
5. The public checkpoint preserves native tensor names without a weight conversion.

A self-contained `PreTrainedModel` / `PretrainedConfig` adapter already exists and has been validated on the released 201.2M SFT checkpoint. It supports modern and legacy KV caches, generation, and `AutoConfig` / `AutoModelForCausalLM` through `auto_map`.

Validation completed:

- exact full-logit parity with the reference PyTorch implementation
- exact cached-step logit parity with the reference cache path
- greedy generation verified
- Transformers 4.30 and 5.14 tested
- 73 focused framework/model tests passed, 3 skipped
- Ruff check and format check passed

Native support would unblock standard evaluation, Intel AutoRound quantization, and downstream serving integrations that reject unknown `model_type` values or remote code.

Expected native loading:

```python
from transformers import AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained(
    "AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT"
)
```

Definition of done:

1. `AutoConfig` recognizes `tr_hash_moe`.
2. `AutoModelForCausalLM` loads the public checkpoint without remote code.
3. Persisted routing tables and stacked 3D expert weights load without conversion.
4. Eager and cached logits match the reference implementation within the agreed tolerance.
5. Greedy generation matches the reference token sequence.
6. The architecture is documented and covered by standard model tests.

We can provide reference logits, routing traces, small fixtures, and any additional architectural details maintainers need.

### Open source status

- [x] The model implementation is available
- [x] The model weights are available

### Provide useful links for the implementation

- Public SFT checkpoint (Apache-2.0): https://huggingface.co/AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT
- Reference implementation: https://github.com/Complexity-ML/complexity-framework
- Validated Transformers adapter: https://github.com/Complexity-ML/complexity-framework/pull/9
- Current serving engine: https://github.com/Complexity-ML/TR-Hash-i64
- Related vLLM model-support request: https://github.com/vllm-project/vllm/issues/53286
- Authors: @Complexity-ML / AETHORIA-AI

**Comment by Complexity-ML** [comment]
### Public checkpoint evaluation update

Since opening this support request, we completed an additional evaluation of the released checkpoint.

**TR-HASH MoE 200M Full SFT** reached:

- ARC-Easy: **57.24%**
- ARC-Challenge: **27.13%**
- Combined ARC: **47.29%**

Combined ARC is the zero-shot micro-average over the complete public splits: 2,376 ARC-Easy and 1,172 ARC-Challenge questions, scored as causal continuations without a chat template.

For context, under the documented comparison protocol:

- GPT-2 Large, 774M: **42.76%**
- Pythia-410M: **41.91%**
- GPT-2 Medium, 355M: **40.08%**
- OPT-350M: **36.33%**

This benchmark does not change the implementation or parity claims in the PR. It shows that the public checkpoint targeted by the native adapter is a functional released model with measurable downstream results, rather than only an architectural test fixture.

Model and evaluation artifacts:
https://huggingface.co/AETHORIA-AI/TR-HASH-MoE-200M-160B-SFT

![Combined ARC comparison](https://github.com/user-attachments/assets/849cfb88-f938-4434-b636-dbd8115954fc)

Native Transformers support would make these evaluations reproducible through standard tooling without `trust_remote_code=True`.
